### PR TITLE
add galaxy_slots as an n_jobs option and other fixes

### PR DESCRIPTION
**Pull request by qiagu**

### CODE REVIEW

File: tools/sklearn/search_model_validation.xml [hunk 1]
```diff
@@ -88,6 +89,7 @@ optimizers = getattr(model_selection, optimizers)
 
 options = params["search_schemes"]["options"]
 options['cv'] = get_cv( options['cv'].strip() )
+options['n_jobs'] = int(sys.argv[2])
```
Comment (bgruening) [inline tools/sklearn/search_model_validation.xml]: try this `options['n_jobs'] = os.environ.get('GALAXY_SLOTS', 1)`

File: tools/sklearn/model_validation.xml [hunk 2]
```diff
@@ -74,6 +75,7 @@ y=y.ravel()
 
 options = params["model_validation_functions"]["options"]
 options['cv'] = get_cv( options['cv'] )
+options['n_jobs'] = int(sys.argv[2])
```
Comment (bgruening) [inline tools/sklearn/model_validation.xml]: try this `options['n_jobs'] = os.environ.get('GALAXY_SLOTS', 1)`

File: tools/sklearn/main_macros.xml [hunk 3]
```diff
@@ -44,9 +44,15 @@ def feature_selector(inputs):
...
-  elif inputs['selected_algorithm'] in ['RFE', 'RFECV']:
-    if 'scoring' in options and (not options['scoring'] or options['scoring'] == 'None'):
+  elif inputs['selected_algorithm'] == 'RFE':
+    estimator=get_estimator(inputs["estimator_selector"])
+    new_selector = selector(estimator, **options)
+
+  elif inputs['selected_algorithm'] == 'RFECV':
+    if not options['scoring'] or options['scoring'] == 'None':
       options['scoring'] = None
+    options['n_jobs'] = int(sys.argv[2])
```
Comment (bgruening) [inline tools/sklearn/main_macros.xml]: try this `options['n_jobs'] = os.environ.get('GALAXY_SLOTS', 1)`

File: tools/sklearn/ensemble.xml [hunk 4]
```diff
@@ -47,6 +47,8 @@ outfile_predict = "$outfile_predict"
 if params["selected_tasks"]["selected_task"] == "train":
     algorithm = params["selected_tasks"]["selected_algorithms"]["selected_algorithm"]
     options = params["selected_tasks"]["selected_algorithms"]["options"]
+    if algorithm in ['RandomForestClassifier', 'RandomForestRegressor']:
+        options['n_jobs'] = int(sys.argv[2])
```
Comment (bgruening) [inline tools/sklearn/ensemble.xml]: try this `options['n_jobs'] = os.environ.get('GALAXY_SLOTS', 1)`

File: tools/sklearn/search_model_validation.xml (hunk 1 above)
Comment (qiagu) [reply tools/sklearn/search_model_validation.xml]: The failure was due to package dependency, but your code seems better. 👍

File: tools/sklearn/main_macros.xml [hunk 5]
```diff
@@ -446,6 +447,10 @@ def get_cv(literal):
     <param argument="fit_intercept" type="boolean" optional="true" truevalue="booltrue" falsevalue="boolfalse" checked="@CHECKED@" label="Estimate the intercept" help="If false, the data is assumed to be already centered."/>
   </xml>
 
+  <xml name="n_jobs" token_default_value="1" token_label="The number of jobs to run in parallel">
+    <param argument="n_jobs" type="integer" min="-1" max="32" value="@DEFAULT_VALUE@" optional="true" label="@LABEL@" help="To use GALAXY_SLOTS, set to 0. Hint: n_jobs > 1 might cause job abortion with no error message."/>
```
Comment (bgruening) [inline tools/sklearn/main_macros.xml]: I would not offer this to the user.

File: tools/sklearn/main_macros.xml [hunk 6]
```diff
@@ -51,7 +51,8 @@ def feature_selector(inputs):
   elif inputs['selected_algorithm'] == 'RFECV':
     if not options['scoring'] or options['scoring'] == 'None':
       options['scoring'] = None
-    options['n_jobs'] = int( os.environ.get('GALAXY_SLOTS', 1) )
+    if options['n_jobs'] == 0:
+      options['n_jobs'] = int( os.environ.get('GALAXY_SLOTS', 1) )
```
Comment (bgruening) [inline tools/sklearn/main_macros.xml]: can we not use
```
slots = os.environ.get('GALAXY_SLOTS', None)
if slots:
   slots = int(slots)
```

File: tools/sklearn/main_macros.xml (hunk 5 above)
Comment (qiagu) [reply tools/sklearn/main_macros.xml]: How about either GALAXY_SLOTS or 1? Users can choose

File: tools/sklearn/main_macros.xml (hunk 6 above)
Comment (qiagu) [reply tools/sklearn/main_macros.xml]: Yes, `None` = 1 here.

File: tools/sklearn/main_macros.xml (hunk 5 above)
Comment (bgruening) [reply tools/sklearn/main_macros.xml]: The user has no change to know how many cores are available. I would not expose this.

Comment (qiagu) [reply tools/sklearn/main_macros.xml]: Okay. I'll remove the change.

File: tools/sklearn/main_macros.xml [hunk 7]
```diff
@@ -1031,6 +1040,16 @@ def get_cv(literal):
     </when>
   </xml>
 
+  <xml name="cv">
+    <param argument="cv" type="text" value="" size="50" optional="true" label="cv" help="Optional. Integer or evalable splitter object, e.g., StratifiedKFold(n_splits=3, shuffle=True, random_state=10). Leave blank for default." >
```
Comment (bgruening) [inline tools/sklearn/main_macros.xml]: the size attribute is deprecated and should be removed from everywhere ...

**Pull request merged**